Tests/PenguinStructuresTests/ArrayStorageExtensionTests.swift

```diff
@@ -129,96 +119,103 @@ extension FactoidArrayStorage where Element == Truthy {
-    XCTAssert(
-      FactoidArrayStorage<Truthy>(minimumCapacity: 0).elementType
-        == Truthy.self)
+    XCTAssert(ArrayStorage<Truthy>.elementType == Truthy.self)
   }
   
   func test_replacementStorage() {
-    FactoidArrayStorage<Truthy>.test_replacementStorage(source: factoids(0..<10))
+    ArrayStorage<Truthy>.test_replacementStorage(source: factoids(0..<10))
   }
   
   func test_makeCopy() {
-    FactoidArrayStorage<Truthy>.test_makeCopy(source: factoids(0..<10))
+    ArrayStorage<Truthy>.test_makeCopy(source: factoids(0..<10))
   }
   
   func test_deinit() {
-    FactoidArrayStorage<Tracked<Truthy>>.test_deinit {
+    ArrayStorage<Tracked<Truthy>>.test_deinit {
       Tracked(Truthy(denominator: 2), track: $0)
     }
   }
 
   func test_copyingInit() {
-    FactoidArrayStorage<Truthy>.test_copyingInit(source: factoids(0..<50))
+    ArrayStorage<Truthy>.test_copyingInit(source: factoids(0..<50))
   }
 
   func test_unsafeInitializingInit() {
-    FactoidArrayStorage<Truthy>.test_unsafeInitializingInit(
+    ArrayStorage<Truthy>.test_unsafeInitializingInit(
       source: factoids(0..<50))
   }
 
   func test_collectionSemantics() {
     let expected = factoids(0..<35)
-    var s = FactoidArrayStorage(expected)
+    var s = ArrayStorage(expected)
     s.checkRandomAccessCollectionSemantics(expectedValues: expected)
     s.checkMutableCollectionSemantics(source: factoids(35..<70))
   }
 
   
   func test_totalError() {
-    FactoidArrayStorage<Truthy>.test_totalError()
+    ArrayStorage<Truthy>.test_totalError()
   }
   
   func test_typeErasedTotalError() {
-    FactoidArrayStorage<Truthy>.test_totalError(typeErased: true)
+    ArrayStorage<Truthy>.test_totalError(typeErased: true)
   }
 
   func test_totalErrorArrayBuffer() {
-    let s = ArrayBuffer<FactoidArrayStorage<Truthy>>(factoids(0..<10))
+    let s = ArrayBuffer<Truthy>(factoids(0..<10))
     let latest = Tracked(0.5) { _ in }
     let total = s.totalError(latest: latest)
     XCTAssertEqual(total, expectedTotalError(0..<10, latest: 0.5))
   }
 
   func test_totalErrorAnyArrayBuffer() {
-    let s: AnyArrayBuffer<FactoidArrayStorage<Truthy>> =
-      AnyArrayBuffer(ArrayBuffer<FactoidArrayStorage<Truthy>>(factoids(0..<10)))
+    let s = AnyArrayBuffer(ArrayBuffer<Truthy>(factoids(0..<10)))
     let latest = Tracked(0.5) { _ in }
-    let total = withUnsafePointer(to: latest) { s.totalError(latestAt: $0) }
+    let total = s.totalError(assumingElementType: Truthy.self, latest: latest)
     XCTAssertEqual(total, expectedTotalError(0..<10, latest: 0.5))
   }
 
+  func test_dynamicType() {
+    let typeErasedBuffers: [AnyArrayBuffer] = [
+      .init(ArrayBuffer<Int>()), .init(ArrayBuffer<Truthy>(factoids(0..<10)))]
+
+    XCTAssert(typeErasedBuffers[0].storage as? PopularityRated == nil)
+    XCTAssert(typeErasedBuffers[1].storage as? PopularityRated != nil)
+    XCTAssertEqual(
+      (typeErasedBuffers[1].storage as? PopularityRated)?.popularity, 35)
```

Comment:
The SwiftFusion case is more complicated than this example. I think this strategy can also work in SwiftFusion, it'll just be a bit more complicated.

We can't make an array of factors conform to `Factor` so we'll need a separate protocol for array of `Factors`:

```
/// Contiguous storage of factors of statically unknown type.
protocol LinearizableFactors {
  func errorVectors(at x: VariableAssignments) -> AllVector
}

extension ArrayStorage: LinearizableFactors where Element: LinearizableFactor {
 ...
}

(typeErasedBuffer.storage as? LinearizableFactors).errorVectors(at: x)
```

I think you'll discover this yourself when you migrate SwiftFusion to use this new stuff, I'm just letting you know beforehand in case it's useful.